gdk4/src/subclass/content_provider.rs

```diff
@@ -0,0 +1,433 @@
+}
+
+unsafe extern "C" fn content_provider_detach_clipboard<T: ContentProviderImpl>(
+    ptr: *mut ffi::GdkContentProvider,
+    clipboard_ptr: *mut ffi::GdkClipboard,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.impl_();
+    let wrap: Borrowed<ContentProvider> = from_glib_borrow(ptr);
+    let clipboard = from_glib_borrow(clipboard_ptr);
+
+    imp.detach_clipboard(wrap.unsafe_cast_ref(), &clipboard)
+}
+
+unsafe extern "C" fn content_provider_formats<T: ContentProviderImpl>(
+    ptr: *mut ffi::GdkContentProvider,
+) -> *mut ffi::GdkContentFormats {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.impl_();
+    let wrap: Borrowed<ContentProvider> = from_glib_borrow(ptr);
+
+    imp.formats(wrap.unsafe_cast_ref()).to_glib_full()
+}
+
+unsafe extern "C" fn content_provider_storable_formats<T: ContentProviderImpl>(
+    ptr: *mut ffi::GdkContentProvider,
+) -> *mut ffi::GdkContentFormats {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.impl_();
+    let wrap: Borrowed<ContentProvider> = from_glib_borrow(ptr);
+
+    imp.storable_formats(wrap.unsafe_cast_ref()).to_glib_full()
+}
+
+unsafe extern "C" fn content_provider_write_mime_type_async<T: ContentProviderImpl>(
+    ptr: *mut ffi::GdkContentProvider,
+    mime_type_ptr: *const libc::c_char,
+    stream_ptr: *mut gio::ffi::GOutputStream,
+    priority: libc::c_int,
+    cancellable_ptr: *mut gio::ffi::GCancellable,
+    callback: gio::ffi::GAsyncReadyCallback,
+    user_data: glib::ffi::gpointer,
+)
+{
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.impl_();
+    let wrap: Borrowed<ContentProvider> = from_glib_borrow(ptr);
+    let cancellable = gio::Cancellable::from_glib_borrow(cancellable_ptr);
+
+    let closure = move |result: &gio::AsyncResult, source_object: Option<&glib::Object>| {
+        let result: *mut gio::ffi::GAsyncResult = result.to_glib_none().0;
+        let source_object: *mut glib::object::GObject = source_object.to_glib_none().0;
+        callback.unwrap()(source_object, result, user_data)
+    };
+
+    let source_object = &ContentProvider::from_glib_borrow(ptr);
+    let t = gio::Task::new(
+        Some(&source_object.upcast_ref::<glib::Object>()),
+        Some(cancellable.as_ref()),
+        closure,
+    );
+
+    glib::MainContext::default().spawn_local(async move {
+        todo!()
+        // imp.write_mime_type_future(wrap.unsafe_cast_ref(),
+        // ).await
+        //     .
+        // task.return_value(true.to_value());
+    });
+}
+
+unsafe extern "C" fn content_provider_write_mime_type_finish<T: ContentProviderImpl>(
+    _ptr: *mut ffi::GdkContentProvider,
+    res_ptr: *mut gio::ffi::GAsyncResult,
+    error_ptr: *mut *mut glib::ffi::GError,
+) -> glib::ffi::gboolean {
+    let res: gio::AsyncResult = from_glib_none(res_ptr);
+    let t = res.downcast_ref::<gio::Task>().unwrap();
+    let ret = t.propagate_value();
```

Comment:
You need to make it so that it either stores a `bool` or a `GError` on errors, so that the implementor can actually return a proper error on failures.

Reply:
yes, should be ok now